plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java

```diff
@@ -0,0 +1,309 @@
+            Optional<JdbcOutputTableHandle> jdbcDeleteOutputTableHandle,
+            JdbcClient jdbcClient,
+            ConnectorPageSinkId pageSinkId,
+            RemoteQueryModifier remoteQueryModifier)
+    {
+        if (jdbcDeleteOutputTableHandle.isPresent()) {
+            return new JdbcPageSink(session, jdbcDeleteOutputTableHandle.get(), jdbcClient, pageSinkId, remoteQueryModifier);
+        }
+
+        JdbcOutputTableHandle deleteOutputTableHandle = new JdbcOutputTableHandle(
+                catalogName,
+                schemaName,
+                tableName,
+                mergeRowIdFieldNames,
+                mergeRowIdFieldTypes,
+                Optional.empty(),
+                Optional.empty(),
+                Optional.empty());
+        return new DeleteSink(session, deleteOutputTableHandle, jdbcClient, pageSinkId, remoteQueryModifier);
+    }
+
+    private class UpdateSink
+            extends JdbcPageSink
+    {
+        public UpdateSink(ConnectorSession session, JdbcOutputTableHandle handle, JdbcClient jdbcClient, ConnectorPageSinkId pageSinkId, RemoteQueryModifier remoteQueryModifier)
+        {
+            super(session, handle, jdbcClient, pageSinkId, remoteQueryModifier);
+        }
+
+        @Override
+        protected String getSinkSql(JdbcClient jdbcClient, JdbcOutputTableHandle outputTableHandle, List<WriteFunction> columnWriters)
+        {
+            List<String> columnNames = outputTableHandle.getColumnNames();
+            checkArgument(columnNames.size() > mergeRowIdFieldNames.size(), "Update primary key itself is not supported");
+            checkArgument(columnNames.size() == columnWriters.size(), "handle and columnWriters mismatch: %s, %s", outputTableHandle, columnWriters);
+
+            ImmutableList.Builder<String> updateConjunctsBuilder = ImmutableList.builder();
+            for (int i = 0; i < columnWriters.size() - mergeRowIdFieldNames.size(); i++) {
+                updateConjunctsBuilder.add(jdbcClient.quoted(columnNames.get(i)) + " = " + columnWriters.get(i).getBindExpression());
+            }
+            return "UPDATE %s.%s SET %s WHERE %s".formatted(jdbcClient.quoted(schemaName), jdbcClient.quoted(tableName), Joiner.on(", ").join(updateConjunctsBuilder.build()), mergeRowIdConjuncts);
+        }
+    }
+
+    private class DeleteSink
+            extends JdbcPageSink
+    {
+        public DeleteSink(ConnectorSession session, JdbcOutputTableHandle handle, JdbcClient jdbcClient, ConnectorPageSinkId pageSinkId, RemoteQueryModifier remoteQueryModifier)
+        {
+            super(session, handle, jdbcClient, pageSinkId, remoteQueryModifier);
+        }
+
+        @Override
+        protected String getSinkSql(JdbcClient jdbcClient, JdbcOutputTableHandle outputTableHandle, List<WriteFunction> columnWriters)
+        {
+            return "DELETE FROM %s.%s WHERE %s".formatted(jdbcClient.quoted(schemaName), jdbcClient.quoted(tableName), mergeRowIdConjuncts);
```

Comment:
I believe we should somehow reuse `io.trino.plugin.jdbc.DefaultQueryBuilder#prepareDeleteQuery`